### PR TITLE
Move file checking into core.

### PR DESCRIPTION
**Pull request by samscott89**
Reduces the duplication needed, and unlocks handling conflicts
in the core (e.g. remove all rules whose source matches the
new filename).

Made the usage of `load` and `load_str` a little more consistent
across core + host languages.

### CODE REVIEW

File: languages/java/oso/src/main/java/com/osohq/oso/Polar.java [hunk 1]
```diff
@@ -2,15 +2,11 @@
```
Comment (samscott89) [inline languages/java/oso/src/main/java/com/osohq/oso/Polar.java:2]: Bunch of autoformat changes here, sorry. Can back out if necessary

File: languages/js/src/Polar.test.ts [hunk 2]
```diff
@@ -358,26 +355,22 @@ describe('#loadFile', () => {
   test('throws if two files with the same contents are loaded', async () => {
     const p = new Polar();
     expect(p.loadFile(await tempFile('', 'a.polar'))).resolves;
-    expect(p.loadFile(await tempFile('', 'b.polar'))).rejects.toThrow(
-      PolarFileDuplicateContentError
-    );
+    expect(p.loadFile(await tempFile('', 'b.polar'))).rejects.toThrow();
```
Comment (samscott89) [inline languages/js/src/Polar.test.ts:358]: @gj couldn't work out the right way to check that the exception thrown has kind `RuntimeError::FileLoading` as specifier in the wasm errors.rs (without manually catching the error and checking it).

File: polar-core/src/polar.rs [hunk 3]
```diff
@@ -70,10 +75,58 @@ impl Polar {
         Self {
             kb: Arc::new(RwLock::new(KnowledgeBase::new())),
             messages: MessageQueue::new(),
+            loaded_content: Arc::new(RwLock::new(HashMap::new())),
+            loaded_files: Arc::new(RwLock::new(HashSet::new())),
         }
     }
 
-    pub fn load_file(&self, src: &str, filename: Option<String>) -> PolarResult<()> {
+    fn check_file(&self, src: &str, filename: &str) -> PolarResult<()> {
```
Comment (samscott89) [inline polar-core/src/polar.rs:83]: Simply relying on the hashing algorithm used in `HashSet` and `HashMap` to check for collisions. This isn't guaranteed to be stable across different versions of rustc. Do we think this is a problem?

File: languages/js/src/Polar.test.ts (hunk 2 above)
Comment (gj) [reply languages/js/src/Polar.test.ts:358]: If you pass a string to `toThrow()` it will do a regex match on the error output

Comment (samscott89) [reply languages/js/src/Polar.test.ts:358]: Ohh, regex. Nice.

File: languages/java/oso/src/main/java/com/osohq/oso/Host.java [hunk 4]
```diff
@@ -92,16 +92,13 @@ public Long cacheInstance(Object instance, Long id) throws Exceptions.OsoExcepti
      * Make an instance of a Java class from a {@code Map<String, Object>} of
      * fields.
      */
-    public Object makeInstance(String className, List<Object> initargs, long id)
-            throws Exceptions.OsoException
-    {
+    public Object makeInstance(String className, List<Object> initargs, long id) throws Exceptions.OsoException {
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Host.java:95]: Are these formatting changes due to version differences in rustfmt?

File: languages/js/src/Polar.test.ts (hunk 2 above)
Comment (samscott89) [reply languages/js/src/Polar.test.ts]: Added a regex test check in the latest.

File: languages/java/oso/src/main/java/com/osohq/oso/Polar.java [hunk 5]
```diff
@@ -100,7 +72,7 @@ public void loadFile(String filename) throws IOException, OsoException {
      * @param filename Name of the source file.
      */
     public void loadStr(String str, String filename) throws Exceptions.OsoException {
```
Comment (leina05) [inline languages/java/oso/src/main/java/com/osohq/oso/Polar.java:74]: I wonder if this method could be private

File: polar-core/src/polar.rs [hunk 6]
```diff
@@ -70,10 +75,58 @@ impl Polar {
         Self {
             kb: Arc::new(RwLock::new(KnowledgeBase::new())),
             messages: MessageQueue::new(),
+            loaded_content: Arc::new(RwLock::new(HashMap::new())),
+            loaded_files: Arc::new(RwLock::new(HashSet::new())),
```
Comment (leina05) [inline polar-core/src/polar.rs:79]:
```suggestion
            loaded_content: Arc::new(RwLock::new(HashMap::new())),    // file content -> file name
            loaded_files: Arc::new(RwLock::new(HashSet::new())),            // set of file names
```

File: polar-core/src/polar.rs [hunk 7]
```diff
@@ -70,10 +75,58 @@ impl Polar {
...
+            .unwrap()
+            .insert(filename.to_string());
+
+        Ok(())
+    }
+
+    pub fn load(&self, src: &str, filename: Option<String>) -> PolarResult<()> {
+        if let Some(ref filename) = filename {
+            self.check_file(src, filename)?;
+        }
```
Comment (leina05) [inline polar-core/src/polar.rs:129]: Way cleaner, this is great

File: polar-core/src/polar.rs (hunk 3 above)
Comment (leina05) [reply polar-core/src/polar.rs:83]: What's the implication of this? These checks won't always work?

Comment (samscott89) [reply polar-core/src/polar.rs:83]: I'm pretty sure its a non-issue. The checks wouldn't match if we somehow made it possible to do hot-swapped oso updates. But even then it's pretty nebulous.

File: languages/java/oso/src/main/java/com/osohq/oso/Host.java (hunk 4 above)
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Host.java:95]: We don't have a standardised java formatter, but apparently my IDE does.

File: polar-core/src/polar.rs (hunk 7 above)
Comment (samscott89) [reply polar-core/src/polar.rs:129]: :)

File: languages/java/oso/src/main/java/com/osohq/oso/Host.java (hunk 4 above)
Comment (gj) [reply languages/java/oso/src/main/java/com/osohq/oso/Host.java:95]: It would be good to pick one and standardize on it (not as part of this PR)

**Pull request merged**